File: examples/cpp/27_collisions.cpp
// Collisions Demo
// This example connects to an RB-Y1 robot, powers on the specified devices,
// monitors collision distance during motion, and sends a stop command when the robot gets too close.
//
// Usage example:
//   ./example_26_collisions --address 192.168.30.1:50051 --model a --power ".*" --servo ".*"
//
// Copyright (c) 2025 Rainbow Robotics. All rights reserved.
//
// DISCLAIMER:
// This is a sample code provided for educational and reference purposes only.
// Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
// the use or misuse of this demo code. Please use with caution and at your own discretion.

#include <algorithm>
#include <atomic>
#include <cctype>
#include <chrono>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <string>
#include <thread>

#include <Eigen/Dense>

#include "00_helper.cpp"
#include "rby1-sdk/model.h"
#include "rby1-sdk/robot.h"

using namespace rb;
using namespace std::chrono_literals;

namespace {

constexpr double kStopDistance = 0.03;
const Eigen::Vector<double, 7> kRightArmTarget{0.8, -0.8, 0.7, -2.0, 0.0, 0.0, 0.0};

std::atomic<bool> g_stop{false};
std::atomic<bool> g_stop_requested{false};

void SignalHandler(int) {
  g_stop = true;
  g_stop_requested = true;
}

std::string ToLower(std::string v) {
  std::transform(v.begin(), v.end(), v.begin(), [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return v;
}

void PrintUsage(const char* prog) {
  std::cerr << "Usage: " << prog
            << " --address <server address> [--model a|m|ub] [--power <regex>] [--servo <regex>]" << std::endl;
  std::cerr << "   or: " << prog << " <server address> [model] [power_regex] [servo_regex]" << std::endl;
}

template <typename ModelT>
bool MoveToPreControlPose(const std::shared_ptr<Robot<ModelT>>& robot) {
  Eigen::VectorXd torso(6);
  Eigen::VectorXd right_arm(7);
  Eigen::VectorXd left_arm(7);
  torso << 0.0, 0.1, -0.2, 0.1, 0.0, 0.0;
  right_arm << 0.2, -0.2, 0.0, -1.0, 0.0, 0.7, 0.0;
  left_arm << 0.2, 0.2, 0.0, -1.0, 0.0, 0.7, 0.0;

  return MoveJ<ModelT>(robot, &torso, &right_arm, &left_arm, 5.0);
}

template <typename ModelT>
int RunRobotCollisions(const std::string& address, const std::string& power_regex, const std::string& servo_regex) {
  auto robot = InitializeRobot<ModelT>(address, power_regex, servo_regex);
  if (!robot) {
    std::cerr << "Failed to initialize robot" << std::endl;
    return 1;
  }

  if (!MoveToPreControlPose(robot)) {
    std::cerr << "Failed to move to pre control pose" << std::endl;
    return 1;
  }

  g_stop_requested = false;
  robot->StartStateUpdate(
      [](const auto& state) {
        if (state.collisions.empty()) {
          return;
        }

        const auto nearest_it = std::min_element(
            state.collisions.begin(), state.collisions.end(),
            [](const auto& a, const auto& b) { return a.distance < b.distance; });
        const auto& nearest = *nearest_it;

        std::cout << std::fixed << std::setprecision(4) << "nearest collision distance: " << nearest.distance << " | "
                  << nearest.link1 << " <-> " << nearest.link2 << std::endl;

        if (nearest.distance < kStopDistance) {
          std::cout << "Collision Detected" << std::endl;
          g_stop_requested = true;
        }
      },
      50.0);

  auto motion_handle = robot->SendCommand(
      RobotCommandBuilder().SetCommand(ComponentBasedCommandBuilder().SetBodyCommand(
          BodyComponentBasedCommandBuilder().SetRightArmCommand(
              JointPositionCommandBuilder().SetMinimumTime(5.0).SetPosition(kRightArmTarget)))),
      10);

  while (!g_stop && !g_stop_requested) {
    std::this_thread::sleep_for(10ms);
  }

  if (g_stop_requested) {
    robot->CancelControl();
  }

  const auto motion_feedback = motion_handle->Get();
  std::cout << "motion finish_code: " << static_cast<int>(motion_feedback.finish_code()) << std::endl;

  robot->StopStateUpdate();

  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  std::signal(SIGINT, SignalHandler);

  std::string address;
  std::string model = "a";
  std::string power = ".*";
  std::string servo = ".*";

  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--address" && i + 1 < argc) {
      address = argv[++i];
    } else if (arg == "--model" && i + 1 < argc) {
      model = argv[++i];
    } else if (arg == "--power" && i + 1 < argc) {
      power = argv[++i];
    } else if (arg == "--servo" && i + 1 < argc) {
      servo = argv[++i];
    } else if (arg.rfind("--", 0) == 0) {
      PrintUsage(argv[0]);
      return 1;
    } else if (address.empty()) {
      address = arg;
    } else if (model == "a") {
      model = arg;
    } else if (power == ".*") {
      power = arg;
    } else if (servo == ".*") {
      servo = arg;
    } else {
      PrintUsage(argv[0]);
      return 1;
    }
  }

  if (address.empty()) {
    PrintUsage(argv[0]);
    return 1;
  }

  model = ToLower(model);

  if (model == "a") {
    return RunRobotCollisions<y1_model::A>(address, power, servo);
  }
  if (model == "m") {
    return RunRobotCollisions<y1_model::M>(address, power, servo);
  }
  if (model == "ub") {
    return RunRobotCollisions<y1_model::UB>(address, power, servo);
  }

  std::cerr << "Unknown model: " << model << std::endl;
  PrintUsage(argv[0]);
  return 1;
}

File: examples/python/27_collisions.py
# Collisions Demo
# This example connects to an RB-Y1 robot, powers on the specified devices,
# monitors collision distance during motion, and sends a stop command when the robot gets too close. See --help for arguments.
#
# Usage example:
#     python 26_collisions.py --address 192.168.30.1:50051 --model a --power ".*"
#
# Copyright (c) 2025 Rainbow Robotics. All rights reserved.
#
# DISCLAIMER:
# This is a sample code provided for educational and reference purposes only.
# Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
# the use or misuse of this demo code. Please use with caution and at your own discretion.

import argparse
import logging
import threading
import time

import numpy as np
import rby1_sdk as rby
import importlib

helper = importlib.import_module("00_helper")
initialize_robot = helper.initialize_robot
movej = helper.movej

STOP_DISTANCE = 0.03
RIGHT_ARM_TARGET = np.array(
    [0.8, -0.8, 0.7, -2.0, 0, 0.0, 0]
    
)
stop_requested = False



def callback(robot_state):
    global stop_requested
    if not robot_state.collisions:
        return

    nearest = min(robot_state.collisions, key=lambda collision: collision.distance)
    print(
        f"nearest collision distance: {nearest.distance:.4f} | "
        f"{nearest.link1} <-> {nearest.link2}"
    )

    if nearest.distance < STOP_DISTANCE:
        print(f"Collision Detected")
        stop_requested = True

def move_to_pre_control_pose(robot):
    """Move to the pre-control pose before starting the motion."""
    torso = np.array([0.0, 0.1, -0.2, 0.1, 0.0, 0.0])
    right_arm = np.array([0.2, -0.2, 0.0, -1.0, 0, 0.7, 0.0])
    left_arm = np.array([0.2, 0.2, 0.0, -1.0, 0, 0.7, 0.0])
    if not movej(robot, torso=torso, right_arm=right_arm, left_arm=left_arm, minimum_time=5.0):
        exit(1)

def main(address, model, power, servo):
    global stop_requested
    robot = initialize_robot(address, model, power, servo)

    move_to_pre_control_pose(robot)

    stop_requested = False
    robot.start_state_update(callback, rate=50)

    rv = movej(robot, right_arm=RIGHT_ARM_TARGET, minimum_time=5.0)

    while not stop_requested:
        time.sleep(0.01)

    
    robot.cancel_control()
    robot.stop_state_update()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="26_collisions")
    parser.add_argument("--address", type=str, required=True, help="Robot address")
    parser.add_argument(
        "--model", type=str, default="a", help="Robot Model Name (default: 'a')"
    )
    parser.add_argument(
        "--power",
        type=str,
        default=".*",
        help="Power device name regex pattern (default: '.*')",
    )
    parser.add_argument(
        "--servo",
        type=str,
        default=".*",
        help="Servo name regex pattern (default: '.*')",
    )
    args = parser.parse_args()

    main(address=args.address, model=args.model, power=args.power, servo=args.servo)
